### PR TITLE
GH-51275: [C++] Allow the CSV Reader to accept multi-character delimiters

### PR DESCRIPTION
**Pull request by HuaHuaY**
### Rationale for this change

Now the type of `delimiter` in `ParseOptions` is `char`, and we can only read csv files whose delimiter is one character. I think we can lift this restriction.

### What changes are included in this PR?

Added a higher-priority `delimiter_string` to `ParseOptions`, and added support for multi-character delimiters in both the chunker and the parser.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes. A new option `delimiter_string` is added to `ParseOptions`.
* GitHub Issue: #51275

### CODE REVIEW

**Comment by HuaHuaY** [comment]
@ursabot please benchmark lang=C++

**Comment by rok** [comment]
Benchmark runs are scheduled for commit 31adcd23be1354c18c45ea2abb689362190915c8. Watch https://buildkite.com/apache-arrow and https://conbench.arrow-dev.org for updates. A comment will be posted here when the runs are complete.

File: cpp/src/arrow/dataset/file_csv.cc [hunk 1]
```diff
@@ -373,6 +373,7 @@ bool CsvFileFormat::Equals(const FileFormat& format) const {
       checked_cast<const CsvFileFormat&>(format).parse_options;
 
   return parse_options.delimiter == other_parse_options.delimiter &&
+         parse_options.delimiter_string == other_parse_options.delimiter_string &&
          parse_options.quoting == other_parse_options.quoting &&
```
[the remark this answers is not in the document]
Comment (HuaHuaY) [reply cpp/src/arrow/dataset/file_csv.cc:377]: I'm not sure whether semantic equality is what we need here.

**Comment by HuaHuaY** [comment]
@ursabot please benchmark lang=C++

**Comment by rok** [comment]
Benchmark runs are scheduled for commit a161efff34c436a89e2120628a67d88f113cf301. Watch https://buildkite.com/apache-arrow and https://conbench.arrow-dev.org for updates. A comment will be posted here when the runs are complete.

File: cpp/src/arrow/csv/options.h [hunk 2]
```diff
@@ -65,6 +65,8 @@ struct ARROW_EXPORT ParseOptions {
   bool pad_short_rows = false;
   /// Whether rows with more columns than expected should ignore the extra columns.
   bool ignore_extra_columns = false;
+  /// Multi-character field delimiter.  If non-empty, overrides `delimiter`.
+  std::string delimiter_string;
```
Comment (wgtmac) [inline cpp/src/arrow/csv/options.h]: Is it a good idea to rename it to `non_compliant_delimiter`? Or do we want to deprecate the above `delimiter` field at all?

File: cpp/src/arrow/csv/lexing_internal.h [hunk 3]
```diff
@@ -61,6 +64,58 @@ decltype(auto) DispatchBool(Fn&& fn, Rest... rest)
...
+  if (!options.delimiter_string.empty()) {
+    return options.delimiter_string.front();
+  }
+  return options.delimiter;
+}
+
+class StreamingDelimiterMatcher {
+ public:
+  explicit StreamingDelimiterMatcher(std::string_view delimiter) : delimiter_(delimiter) {
+    DCHECK(!delimiter_.empty());
```
Comment (wgtmac) [inline cpp/src/arrow/csv/lexing_internal.h]: nit: add an error msg

File: cpp/src/arrow/dataset/file_csv.cc [hunk 4]
```diff
@@ -373,6 +373,7 @@ bool CsvFileFormat::Equals(const FileFormat& format) const {
       checked_cast<const CsvFileFormat&>(format).parse_options;
 
   return parse_options.delimiter == other_parse_options.delimiter &&
```
Comment (wgtmac) [inline cpp/src/arrow/dataset/file_csv.cc]: Because `delimiter_string` overrides `delimiter` when it is non-empty, comparing both fields makes two formats look different even when they parse exactly the same way. For example, `delimiter=';', delimiter_string="||"` and `delimiter=',', delimiter_string="||"` behave the same, but this returns false. Since `FileFragment::Equals()` uses this, it can affect dataset matching/caching. Could we compare the effective delimiter instead and add a small regression test?

File: cpp/src/arrow/csv/lexing_internal.h [hunk 5]
```diff
@@ -61,6 +64,58 @@ decltype(auto) DispatchBool(Fn&& fn, Rest... rest)
...
+    if (c == delimiter_[matched_]) {
+      matched_++;
+      if (matched_ == delimiter_.size()) {
+        matched_ = 0;
+        return true;
+      }
+      return false;
+    }
+
+    const std::string_view delimiter(delimiter_);
```
Comment (wgtmac) [inline cpp/src/arrow/csv/lexing_internal.h]: This fallback logic is the tricky part of the new matcher, but the added tests only cover the case where a partial match becomes a complete match. Could you add tests where a partial match is followed by a mismatch, and where the delimiter has an overlapping prefix, such as `aba` or `abab`? Please also add a parser/chunker case with `quoting=false` so that path is covered.

File: cpp/src/arrow/csv/options.cc [hunk 6]
```diff
@@ -23,8 +23,39 @@ namespace csv {
 ParseOptions ParseOptions::Defaults() { return ParseOptions(); }
 
 Status ParseOptions::Validate() const {
-  if (ARROW_PREDICT_FALSE(delimiter == '\n' || delimiter == '\r')) {
-    return Status::Invalid("ParseOptions: delimiter cannot be \\r or \\n");
+  // The chunker handles escapes before delimiter matching, so allowing the escape
+  // character in a delimiter could make it disagree with the parser.
+  if (escaping) {
```
Comment (wgtmac) [inline cpp/src/arrow/csv/options.cc:28]: These are new (benign) behavior changes. I think we need to add some test cases to verify that these malformed inputs are now explicitly rejected.

File: cpp/src/arrow/dataset/file_csv.cc [hunk 7]
```diff
@@ -499,6 +500,10 @@ Result<std::shared_ptr<FileWriter>> CsvFileFormat::MakeWriter(
   if (!Equals(*options->format())) {
     return Status::TypeError("Mismatching format/write options.");
   }
+  if (!parse_options.delimiter_string.empty()) {
```
Comment (wgtmac) [inline cpp/src/arrow/dataset/file_csv.cc:503]: The new Dataset behavior has no tests. Could you add one for format equality with the same effective delimiter but different legacy `delimiter`?

File: cpp/src/arrow/csv/options.cc [hunk 8]
```diff
@@ -23,8 +23,39 @@ namespace csv {
 ParseOptions ParseOptions::Defaults() { return ParseOptions(); }
 
 Status ParseOptions::Validate() const {
-  if (ARROW_PREDICT_FALSE(delimiter == '\n' || delimiter == '\r')) {
-    return Status::Invalid("ParseOptions: delimiter cannot be \\r or \\n");
+  // The chunker handles escapes before delimiter matching, so allowing the escape
```
Comment (wgtmac) [inline cpp/src/arrow/csv/options.cc:26]: Is it cleaner to consolidate the two cases like below?

```cpp
  std::string invalid_chars = "\r\n";
  if (escaping) {
    invalid_chars.push_back(escape_char);
  }

  const bool use_delimiter_string = !delimiter_string.empty();
  const auto invalid_pos =
      use_delimiter_string ? delimiter_string.find_first_of(invalid_chars)
                           : invalid_chars.find(delimiter);
  if (invalid_pos != std::string::npos) {
    return Status::Invalid("ParseOptions: delimiter contains an invalid character");
  }

  const char delimiter_first_byte =
      use_delimiter_string ? delimiter_string.front() : delimiter;
  if (quoting && delimiter_first_byte == quote_char) {
    return Status::Invalid(
        "ParseOptions: delimiter cannot start with the quote character");
  }
```

File: cpp/src/arrow/csv/lexing_internal.h (hunk 5 above)
Comment (wgtmac) [reply cpp/src/arrow/csv/lexing_internal.h]: BTW, this piece of logic is a little bit complicated. Perhaps add a line of comment to help understand it.

File: cpp/src/arrow/csv/parser.cc [hunk 9]
```diff
@@ -59,6 +60,22 @@ Status MismatchingColumns(const InvalidRow& row) {
 
 inline bool IsControlChar(uint8_t c) { return c < ' '; }
 
+enum class DelimiterMatch { NoMatch, Match, Incomplete };
```
Comment (wgtmac) [inline cpp/src/arrow/csv/parser.cc:63]: `Incomplete` -> `Partial`? Incomplete seems like failed to match.

File: cpp/src/arrow/csv/parser_test.cc [hunk 10]
```diff
@@ -268,6 +268,38 @@ TEST(BlockParser, Basics) {
...
+  auto options = ParseOptions::Defaults();
+  options.delimiter_string = "||";
+
+  BlockParser parser(options, /*num_cols=*/2);
+  AssertParsePartial(parser, "a||b|", 0);
+  AssertParseFinal(parser, "a||b|");
+  AssertColumnsEq(parser, {{"a"}, {"b|"}});
+}
+
+TEST(BlockParser, SingleCharacterDelimiterString) {
```
Comment (wgtmac) [inline cpp/src/arrow/csv/parser_test.cc:293]: Can we add a few more parser cases here? The current tests cover the happy path, but not simple edge cases like consecutive delimiters (`a||||b`), a delimiter at the start or end of a row, `escaping=true`, or combinations with `pad_short_rows`/`ignore_extra_columns`. These are small tests and exercise the new `FieldStart`/`InField` branches.

File: cpp/src/arrow/dataset/file_csv.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (wgtmac) [reply cpp/src/arrow/dataset/file_csv.cc:377]: I think this is worth doing to make it more user friendly.

File: cpp/src/arrow/dataset/file_csv.cc [hunk 11]
```diff
@@ -499,6 +500,10 @@ Result<std::shared_ptr<FileWriter>> CsvFileFormat::MakeWriter(
...
+    return Status::NotImplemented(
+        "Writing CSV files with delimiter_string is not supported");
```
Comment (wgtmac) [inline cpp/src/arrow/dataset/file_csv.cc:505]: Perhaps letting users know that it is non spec-compliant?

File: cpp/src/arrow/csv/chunker_test.cc [hunk 12]
```diff
@@ -193,6 +193,27 @@ TEST_P(BaseChunkerTest, QuotingNewline) {
...
 }
 
+TEST_P(BaseChunkerTest, MultiDelimiter) {
+  if (!options_.newlines_in_values) {
+    return;
+  }
+  options_.delimiter_string = "||";
+  MakeChunker();
+
+  auto partial = std::make_shared<Buffer>("name|");
```
Comment (wgtmac) [inline cpp/src/arrow/csv/chunker_test.cc:203]: Can we also add a mismatch case here? For example, with `delimiter_string = "||"`, `newlines_in_values = true`, and input `a|b||"c\n d"||e\n`, the single `|` in `a|b` should stay literal, the later `||` should start a quoted field, and the embedded newline should not end the row. This would cover fallback and quote recognition together.
